### PR TITLE
Adding properties to encass entities

### LINKED ISSUE #150: Exported encapsulated assertions loose properties
**Describe the bug**
Encapsulated assertions have properties. These include information such as 'allowTracing', 'paletteIconResourceName'. These should be preserved.

**To Reproduce**
1. Create an encapsulated assertion and set a non-default icon (not the star)
2. Export the encapsulated assertion/ the new policy
3. import that policy into a new gateway

**Expected behavior**
The selected icon should be visible when using the encapsulated assertion in policy. Right now it is not.

**Logs**
---

**Environment (please complete the following information):**
 - Gateway Version 9.3, 9.4

**Additional context**
Overall, no properties should be lost

### PR DESCRIPTION
**Pull request by yves-chan**
Fixes #150 

## Proposed Changes
* Adding properties to encass entities
* updated unit tests

### CODE REVIEW

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/EncassEntityBuilder.java [hunk 1]
```diff
@@ -72,7 +72,11 @@ private Entity buildEncassEntity(Bundle bundle, String name, Encass encass, Docu
                 buildResults(encass, document)
         );
 
-        buildAndAppendPropertiesElement(ImmutableMap.of(PALETTE_FOLDER, INTERNAL_ASSERTIONS), document, encassAssertionElement);
+        buildAndAppendPropertiesElement(
+                encass.getProperties() == null ?
+                        ImmutableMap.of(PALETTE_FOLDER, INTERNAL_ASSERTIONS) :
```
Comment (joao-borges) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/EncassEntityBuilder.java]: I believe you have to check if this property is present in a possible existing map and add if not.

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/loader/EncassLoader.java [hunk 2]
```diff
@@ -31,10 +33,14 @@ public void load(Bundle bundle, Element element) {
         final String policyId = getSingleChildElementAttribute(encassElement, POLICY_REFERENCE, ATTRIBUTE_ID);
         final String name = getSingleChildElementTextContent(encassElement, NAME);
         final String guid = getSingleChildElementTextContent(encassElement, GUID);
+        final Map<String, Object> properties = mapPropertiesElements(getSingleChildElement(encassElement, PROPERTIES, true), PROPERTIES);
+        //Removing redundant policyGuid, since encass already has a reference to policyId
+        properties.remove("policyGuid");
```
Comment (joao-borges) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/loader/EncassLoader.java]: use some constant here or create if that does not exist

File: config-builder/src/test/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/EncassEntityBuilderTest.java [hunk 3]
```diff
@@ -99,6 +97,12 @@ private static void buildBundleWithEncass(Bundle bundle, BundleType deployment,
         assertEquals(encassGuid, getSingleChildElementTextContent(xml, GUID));
         assertNotNull(getSingleChildElement(xml, POLICY_REFERENCE));
         assertEquals(policyId, getSingleChildElement(xml, POLICY_REFERENCE).getAttribute(ATTRIBUTE_ID));
+        assertNotNull(getSingleChildElement(xml, PROPERTIES));
+        Map<String, Object> props = mapPropertiesElements(getSingleChildElement(xml, PROPERTIES), PROPERTIES);
+        assertEquals(3, props.size());
+        assertEquals("internalAssertions", props.get("paletteFolder"));
```
Comment (joao-borges) [inline config-builder/src/test/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/EncassEntityBuilderTest.java]: add constants for the properties names

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/EncassEntityBuilder.java [hunk 4]
```diff
@@ -18,21 +18,22 @@
...
 import static com.ca.apim.gateway.cagatewayconfig.util.entity.EntityTypes.ENCAPSULATED_ASSERTION_TYPE;
 import static com.ca.apim.gateway.cagatewayconfig.util.gateway.BuilderUtils.buildAndAppendPropertiesElement;
 import static com.ca.apim.gateway.cagatewayconfig.util.gateway.BundleElementNames.*;
+import static com.ca.apim.gateway.cagatewayconfig.util.properties.PropertyConstants.PALETTE_FOLDER;
 import static com.ca.apim.gateway.cagatewayconfig.util.xml.DocumentUtils.*;
 import static java.lang.Boolean.FALSE;
 import static org.apache.commons.lang3.ObjectUtils.firstNonNull;
 
 @Singleton
 public class EncassEntityBuilder implements EntityBuilder {
```
Comment (joao-borges) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/EncassEntityBuilder.java]: this should be on PropertyConstants class

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/EncassEntityBuilder.java [hunk 5]
```diff
@@ -72,7 +73,13 @@ private Entity buildEncassEntity(Bundle bundle, String name, Encass encass, Docu
                 buildResults(encass, document)
         );
 
-        buildAndAppendPropertiesElement(ImmutableMap.of(PALETTE_FOLDER, INTERNAL_ASSERTIONS), document, encassAssertionElement);
+        Map<String, Object> properties = encass.getProperties();
```
Comment (joao-borges) [inline config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/EncassEntityBuilder.java]: do something here using the `Optional` class:
`Optional.ofNullable(encass.getProperties()).orElse(new HashMap)`

Then you don't need the if/else below

File: config-builder/src/test/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/EncassEntityBuilderTest.java [hunk 6]
```diff
@@ -151,6 +157,10 @@ private static Encass buildTestEncass(String encassGuid, String policyPath) {
         result2.setType("message");
         encass.getResults().add(result1);
         encass.getResults().add(result2);
+        encass.setProperties(ImmutableMap.of(
+                PALETTE_FOLDER, DEFAULT_PALETTE_FOLDER_LOCATION,
+                "paletteIconResourceName", "someImage",
```
Comment (joao-borges) [inline config-builder/src/test/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/EncassEntityBuilderTest.java]: shouldn't this constants be added to PropertyConstants ?

File: config-builder/src/test/java/com/ca/apim/gateway/cagatewayconfig/bundle/loader/EncassLoaderTest.java [hunk 7]
```diff
@@ -89,6 +90,8 @@ private void load(Bundle bundle) {
         EncassResult result = entity.getResults().iterator().next();
         assertEquals("result", result.getName());
         assertEquals("string", result.getType());
+        assertEquals("internalAssertions", entity.getProperties().get("paletteFolder"));
```
Comment (joao-borges) [inline config-builder/src/test/java/com/ca/apim/gateway/cagatewayconfig/bundle/loader/EncassLoaderTest.java]: Also PropertyConstants?

File: config-builder/src/test/java/com/ca/apim/gateway/cagatewayconfig/config/loader/EncassLoaderTest.java [hunk 8]
```diff
@@ -179,6 +192,12 @@ private static void check(Bundle bundle) {
         assertEquals(2, encass.getResults().size());
         encass.getResults().forEach(e -> assertTrue((e.getName().equals("goodbye") && e.getType().equals("string")) ||
                 (e.getName().equals("goodbye-again") && e.getType().equals("message"))));
+        assertEquals(5, encass.getProperties().size());
+        assertEquals("policyLogic", encass.getProperties().get("paletteFolder"));
```
Comment (joao-borges) [inline config-builder/src/test/java/com/ca/apim/gateway/cagatewayconfig/config/loader/EncassLoaderTest.java]: and also PropertyConstants too?

File: config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/EncassEntityBuilder.java (hunk 4 above)
Comment (yves-chan) [reply config-builder/src/main/java/com/ca/apim/gateway/cagatewayconfig/bundle/builder/EncassEntityBuilder.java]: Moved to PropertyConstants class

**Pull request merged**